File: CS50_week1/week1_pset/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int getLength(long number);
int numIterations(long number);
bool isValid(long number, int iterations);
int sumDigitsFromSecondToLastNumber(long number, int iterations);
int sumDigitsFromLastNumber(long number, int iterations);
string whatCard(long number, int iterations);

int main(void)
{
    long number = get_long("Number: ");
    int iterations = numIterations(number);
    printf("%s\n", whatCard(number, iterations));
}

int getLength(long number)
{
    int count = 0;
    do
    {
        number /= 10;
        count++;
    }
    while (number != 0);
    return count;
}

int numIterations(long number)
{
    int iterations;
    int length = getLength(number);
    if (length % 2 == 0)
    {
        iterations = (length / 2);
    }
    else
    {
        iterations = (length / 2) + 1;
    }
    return iterations;
}

bool isValid(long number, int iterations)
{
    int firstSum = sumDigitsFromSecondToLastNumber(number, iterations);
    int secondSum = sumDigitsFromLastNumber(number, iterations);
    return ((firstSum + secondSum) % 10 == 0);
}

int sumDigitsFromSecondToLastNumber(long number, int iterations)
{
    int sum = 0;
    long n = number / 10;
    for (int i = 0; i < iterations; i++)
    {
        int digit = (n % 10) * 2;
        int digit_length = getLength(digit);
        if (digit_length == 1)
        {
            sum += digit;
        }
        else
        {
            for (int j = 0; j < digit_length; j++)
            {
                int digit2 = (digit % 10);
                digit /= 10;
                sum += digit2;
            }
        }
        n /= 100;
    }
    return sum;
}

int sumDigitsFromLastNumber(long number, int iterations)
{
    int sum = 0;
    long n2 = number;
    for (int i = 0; i < iterations; i++)
    {
        int digit = n2 % 10;
        sum += digit;
        n2 /= 100;
    }
    return sum;
}

string whatCard(long number, int iterations)
{
    long placeholder = number;
    int numLength = getLength(number);
    for (int i = 0; i < numLength - 2; i++)
    {
        placeholder /= 10;
    }
    if (isValid(number, iterations))
    {
        if (placeholder / 10 == 4 && (numLength == 16 || numLength == 13))
        {
            return ("VISA");
        }
        else if ((placeholder == 51 || placeholder == 52 || placeholder == 53 ||
                  placeholder == 54 || placeholder == 55) &&
                 numLength == 16)
        {
            return ("MASTERCARD");
        }
        else if ((placeholder == 34 || placeholder == 37) && numLength == 15)
        {
            return ("AMEX");
        }
        else
        {
            return ("INVALID");
        }
    }
    else
    {
        return ("INVALID");
    }
}

File: CS50_week6/week6pset/sentimental-credit/credit.py
from cs50 import get_int
import math

AMEX_START = [34, 37]
MASTERCARD_START = [51, 52, 53, 54, 55]
VISA_START = 4

AMEX_LENGTH = 15
MASTERCARD_LENGTH = 16
VISA_LENGTH = [13, 16]


def main():
    number = get_int("Number: ")
    print(whatCard(number))


def sumDigitsFromSecondToLastDigit(number, iterations):
    sum = 0
    numcopy = int(number / 10)
    for i in range(iterations):
        n = (numcopy % 10) * 2
        if (numlen(n) == 1):
            sum += n
        else:
            ncopy = str(n)
            for i in range(len(ncopy)):
                sum += int(ncopy[i])
        numcopy = int(numcopy / 100)
    return sum


def sumDigitsFromLastDigit(number, iterations):
    sum = 0
    numcopy2 = number
    for i in range(iterations):
        n = numcopy2 % 10
        sum += n
        numcopy2 = int(numcopy2 / 100)
    return sum


def getFirstTwoDigits(number):
    digits = ""
    n = str(number)
    for i in range(2):
        digits += n[i]
    return digits


def numlen(n):
    numlength = len(str(n))
    return numlength


def isValid(number):
    iterations = math.ceil(numlen(number) / 2)
    checksum = str(sumDigitsFromSecondToLastDigit(number, iterations) +
                   sumDigitsFromLastDigit(number, iterations))
    if int(checksum) % 10 == 0:
        return True
    return False


def whatCard(number):
    digits = getFirstTwoDigits(number)
    if isValid(number):
        if (int(digits[0]) == VISA_START) and (numlen(number) in VISA_LENGTH):
            return "VISA"
        elif (int(digits) in MASTERCARD_START) and (numlen(number) == MASTERCARD_LENGTH):
            return "MASTERCARD"
        elif (int(digits) in AMEX_START) and (numlen(number) == AMEX_LENGTH):
            return "AMEX"
        else:
            return "INVALID"
    else:
        return "INVALID"


main()
